src/hotspot/share/runtime/handshake.cpp

```diff
@@ -701,6 +701,10 @@ void HandshakeState::do_self_suspend() {
   assert(!_handshakee->has_last_Java_frame() || _handshakee->frame_anchor()->walkable(), "should have walkable stack");
   assert(_handshakee->thread_state() == _thread_blocked, "Caller should have transitioned to _thread_blocked");
 
+  // Separate all the writes above for other threads reading state
+  // set by this suspended thread.
+  OrderAccess::fence();
```

Comment:
IIUC what @robehn is concerned about is stuff that happens before we even take the lock, being reordered with the state change to `_thread_blocked` which happens in `do_thread` before this is called, so this placement is far too late to prevent that reordering.  If a fence is needed then it would be just before (or inside) `set_thread_state`.

Note `set_thread_state` already has release semantics on some platforms (obviously there is disagreement about the need for, or placement of, this). And `set_thread_state_fence`, has the fence in the wrong place for the current problem. We seem to want a `fence_and_set_thread_state`.